pallets/rewards/src/lib.rs

```diff
@@ -0,0 +1,410 @@
+			+ EnsureAdd
+			+ EnsureSub;
+
+		/// Type used to handle rates as fixed points numbers.
+		type Rate: FixedPointNumber + TypeInfo + MaxEncodedLen + Encode + Decode;
+
+		/// Type used to identify groups.
+		type GroupId: codec::FullCodec
+			+ TypeInfo
+			+ MaxEncodedLen
+			+ Copy
+			+ PartialEq
+			+ sp_std::fmt::Debug;
+
+		/// Max number of currency movements. See [`Rewards::attach_currency()`].
+		#[pallet::constant]
+		type MaxCurrencyMovements: Get<u32> + TypeInfo;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	// --------------------------
+	//          Storage
+	// --------------------------
+
+	#[pallet::storage]
+	pub(super) type Currencies<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		T::CurrencyId,
+		CurrencyInfo<T::Balance, T::Rate, T::GroupId, T::MaxCurrencyMovements>,
+		ValueQuery,
+	>;
+
+	#[pallet::storage]
+	pub(super) type Groups<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::GroupId, Group<T::Balance, T::Rate>, ValueQuery>;
+
+	#[pallet::storage]
+	pub(super) type StakeAccounts<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::AccountId,
+		Blake2_128Concat,
+		T::CurrencyId,
+		StakeAccount<T::Balance, T::SignedBalance>,
+		ValueQuery,
+	>;
+
+	// --------------------------
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)] // TODO
+	pub enum Event<T: Config> {
+		GroupRewarded {
+			group_id: T::GroupId,
+			amount: T::Balance,
+		},
+		StakeDeposited {
+			group_id: T::GroupId,
+			currency_id: T::CurrencyId,
+			account_id: T::AccountId,
+			amount: T::Balance,
+		},
+		StakeWithdrawn {
+			group_id: T::GroupId,
+			currency_id: T::CurrencyId,
+			account_id: T::AccountId,
+			amount: T::Balance,
+		},
+		RewardClaimed {
+			group_id: T::GroupId,
+			currency_id: T::CurrencyId,
+			account_id: T::AccountId,
+			amount: T::Balance,
+		},
+		CurrencyAttached {
+			currency_id: T::CurrencyId,
+			from: Option<T::GroupId>,
+			to: T::GroupId,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		// Emits when a currency is used but it does not have a group associated to.
+		CurrencyWithoutGroup,
+
+		// Emits when a currency is attached to the group it is already attached.
+		CurrencyInSameGroup,
+
+		// Emits when a currency is moved more than `MaxCurrencyMovements` times.
+		CurrencyMaxMovementsReached,
+	}
+
+	impl<T: Config> GroupRewards for Pallet<T>
+	where
+		T::Balance: EnsureAdd + EnsureSub,
+	{
+		type Balance = T::Balance;
+		type GroupId = T::GroupId;
+
+		fn reward_group(group_id: Self::GroupId, reward: Self::Balance) -> DispatchResult {
+			Groups::<T>::try_mutate(group_id, |group| {
+				group.distribute_reward(reward)?;
+
+				T::Currency::mint_into(
+					T::RewardCurrency::get(),
+					&T::PalletId::get().into_account_truncating(),
+					reward,
+				)?;
+
+				Self::deposit_event(Event::GroupRewarded {
+					group_id,
+					amount: reward,
+				});
+
+				Ok(())
+			})
+		}
```

Comment:
@lemunozm instead of default minting here, we could also refer to some external logic. 

```rust
                 type Balance = T::Balance;
		type GroupId = T::GroupId;
                 type Allocation = T::RewardAllocation;

		fn reward_group(group_id: Self::GroupId, reward: Self::Balance) -> DispatchResult {
			Groups::<T>::try_mutate(group_id, |group| {
				group.distribute_reward(reward)?;

				Allocation::allocate_for(
					group_id,
					&T::PalletId::get().into_account_truncating(),
					reward,
				)?;

				Self::deposit_event(Event::GroupRewarded {
					group_id,
					amount: reward,
				});

				Ok(())
			})
		}
```

where `T::RewardAllocation<T::RewardCurrency>` us bounded by a trait of `trait Allocation<Currency: Get<Self::CurrencyId>>`. We can then let the runtime decide if we wanna mint for some groups or use the pot for some others.

Reply:
I've talked to @Miguel about this. Seems like by now, we will use the `PalletId` for minting inside the pallet. But I will keep this idea in mind to implement later if needed. Thanks @mustermeiszer for the hint!